File: cashboard-backend/core/utils/digikala_scraper.py
# core/utils/digikala_scraper.py
import re
from django.utils import timezone
from digikala_api import product_detail

def fetch_product_info(product_url: str) -> dict:
    """
    Accepts a full Digikala product URL (or a URL-like string).
    Returns a dict with keys: title, price, image_url, updated_at
    On error returns: {"error": "message"}
    """
    if not product_url or not isinstance(product_url, str):
        return {"error": "Invalid product_url"}

    # try to extract a product id (last sequence of digits in the url)
    try:
        matches = re.findall(r'(\d+)', product_url)
        if not matches:
            return {"error": "Could not find product id in the URL"}

        product_id = matches[-1]

        # call digikala api
        product = product_detail(product_id)
        if not product:
            return {"error": "Product not found via digikala_api"}

        title = getattr(product, "persian_name", None) or getattr(product, "english_name", "") or ""
        price = getattr(product, "price", None)
        image_url = getattr(product, "image_url", None)

        return {
            "title": title,
            "price": price,
            "image_url": image_url,
            "updated_at": timezone.now()
        }
    except Exception as e:
        # keep error message simple for now; you can log the exception where appropriate
        return {"error": str(e)}

File: cashboard-frontend/src/utils/digikalaScraper.js
// src/utils/digikalaScraper.js
import axios from 'axios';

/**
 * Fetch product info from Digikala URL
 * @param {string} productUrl - full Digikala product URL
 * @returns {Promise<{title: string, price: number, image_url: string, updated_at: string} | {error: string}>}
 */
export async function fetchProductInfo(productUrl) {
  if (!productUrl || typeof productUrl !== 'string') {
    return { error: 'Invalid URL' };
  }

  try {
    // Extract the last numeric sequence as productId
    const matches = productUrl.match(/(\d+)/g);
    if (!matches) return { error: 'Could not extract product ID from URL' };

    const productId = matches[matches.length - 1];

    // API call with browser-like headers
    const response = await axios.get(`https://api.digikala.com/v1/product/${productId}/`, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)',
        'Accept': 'application/json',
        'Referer': 'https://www.digikala.com/',
      },
      timeout: 5000,
    });

    const product = response.data?.data?.product;

    if (!product) return { error: 'Product not found' };

    const title = product?.title || product?.persian_name || '';
    const price = product?.price?.selling_price || 0;
    const image_url = product?.media?.images?.[0]?.url || '';

    return {
      title,
      price,
      image_url,
      updated_at: new Date().toISOString(),
    };
  } catch (err) {
    console.error('Digikala fetch error:', err.response?.status, err.message);
    return { error: 'Failed to fetch product info' };
  }
}
